In [1]:
import pandas as pd

In [65]:
path = r"https://raw.githubusercontent.com/sindhura-nk/Datasets/refs/heads/main/Salary_dataset.csv"
salary = pd.read_csv(path)
salary

Unnamed: 0  YearsExperience    Salary
0            0              1.2   39344.0
1            1              1.4   46206.0
2            2              1.6   37732.0
3            3              2.1   43526.0
4            4              2.3   39892.0
5            5              3.0   56643.0
6            6              3.1   60151.0
7            7              3.3   54446.0
8            8              3.3   64446.0
9            9              3.8   57190.0
10          10              4.0   63219.0
11          11              4.1   55795.0
12          12              4.1   56958.0
13          13              4.2   57082.0
14          14              4.6   61112.0
15          15              5.0   67939.0
16          16              5.2   66030.0
17          17              5.4   83089.0
18          18              6.0   81364.0
19          19              6.1   93941.0
20          20              6.9   91739.0
21          21              7.2   98274.0
22          22              8.0  101303.0
23          23              8.3  113813.0
24          24              8.8  109432.0
25          25              9.1  105583.0
26          26              9.6  116970.0
27          27              9.7  112636.0
28          28             10.4  122392.0
29          29             10.6  121873.0

In [4]:
salary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       30 non-null     int64  
 1   YearsExperience  30 non-null     float64
 2   Salary           30 non-null     float64
dtypes: float64(2), int64(1)
memory usage: 852.0 bytes


# data Cleaning

In [5]:
salary.isna().sum()

Unnamed: 0         0
YearsExperience    0
Salary             0
dtype: int64

In [7]:
salary.duplicated().sum()

np.int64(0)

# Separate X and Y features
    X Indepandent feature -> YearExperience
    Y Dependent feature -> Salary

In [14]:
salary[['YearsExperience']]

YearsExperience
0               1.2
1               1.4
2               1.6
3               2.1
4               2.3
5               3.0
6               3.1
7               3.3
8               3.3
9               3.8
10              4.0
11              4.1
12              4.1
13              4.2
14              4.6
15              5.0
16              5.2
17              5.4
18              6.0
19              6.1
20              6.9
21              7.2
22              8.0
23              8.3
24              8.8
25              9.1
26              9.6
27              9.7
28             10.4
29             10.6

In [15]:
X = salary[['YearsExperience']]
Y = salary[['Salary']]

In [16]:
X.head()

YearsExperience
0              1.2
1              1.4
2              1.6
3              2.1
4              2.3

In [17]:
Y.head()

Salary
0  39344.0
1  46206.0
2  37732.0
3  43526.0
4  39892.0

# There are no missing value nor any duplicated rows in this dataset

In [18]:
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer

In [19]:
pipe = make_pipeline(SimpleImputer(strategy='median')).set_output(transform='pandas')

In [20]:
pipe

Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='median'))])

In [28]:
X_pre = pipe.fit_transform(X)
X_pre.head()

YearsExperience
0              1.2
1              1.4
2              1.6
3              2.1
4              2.3

# Model Building = Linear Regression
    Equation of Line: y = B0 + B1.X
    B0 => y intercept
    B1 => Slope of the line

    ProfitPredicted = B0 + B1*RND

In [36]:
from sklearn.linear_model import LinearRegression

In [37]:
model = LinearRegression()
model.fit(X,Y)


LinearRegression()

In [38]:
# B0 - y intercept
model.intercept_

array([24848.20396652])

In [39]:
# B1 - slope of the line
model.coef_

array([[9449.96232146]])

salarypredicted = 24848.20 + 9449.96*Salary

Salary = 0 salarypredicted = 24848.20

Salary = 1 it will increase by 9449.96 

In [40]:
Y.head()

Salary
0  39344.0
1  46206.0
2  37732.0
3  43526.0
4  39892.0

In [42]:
Ypreds = model.predict(X)
Ypreds

array([[ 36188.15875227],
       [ 38078.15121656],
       [ 39968.14368085],
       [ 44693.12484158],
       [ 46583.11730587],
       [ 53198.09093089],
       [ 54143.08716303],
       [ 56033.07962732],
       [ 56033.07962732],
       [ 60758.06078805],
       [ 62648.05325234],
       [ 63593.04948449],
       [ 63593.04948449],
       [ 64538.04571663],
       [ 68318.03064522],
       [ 72098.0155738 ],
       [ 73988.00803809],
       [ 75878.00050238],
       [ 81547.97789525],
       [ 82492.9741274 ],
       [ 90052.94398456],
       [ 92887.932681  ],
       [100447.90253816],
       [103282.8912346 ],
       [108007.87239533],
       [110842.86109176],
       [115567.84225249],
       [116512.83848464],
       [123127.81210966],
       [125017.80457395]])

In [43]:
Ypreds[:5]

array([[36188.15875227],
       [38078.15121656],
       [39968.14368085],
       [44693.12484158],
       [46583.11730587]])

In [44]:
Y.head()

Salary
0  39344.0
1  46206.0
2  37732.0
3  43526.0
4  39892.0

# Evaluation Metrics
    Mean Squared Error
    Mean Absolute Error
    RMSE
    R2 squared- R2 score

In [46]:
from sklearn.metrics import mean_squared_error,mean_absolute_error,r2_score

In [47]:
# Mean squared erro
MSE = mean_squared_error(Y, Ypreds)
# RMSE Root mean squared error
RMSE = MSE**(1/2)
# MAE
MAE = mean_absolute_error(Y, Ypreds)
# r2 score
r2 = r2_score(Y, Ypreds)

print(f"Mean Squared Error: {MSE:.2f}")
print(f"Root Mean Squared Error: {RMSE:.2f}")
print(f"Mean Absolute Error: {MAE:.2f}")
print(f"R2 squared: {r2*100:.2f}%")


Mean Squared Error: 31270951.72
Root Mean Squared Error: 5592.04
Mean Absolute Error: 4644.20
R2 squared: 95.70%


# R2 squared value is greater than 80%, we can consider this model for final model building

In [66]:
exp_sample = [0,3.4,7.5,11.1,10.5,5,0.5,2.1]

In [67]:
preds = []

for x in exp_sample:
    pred = model.predict([[x]]).round(2)
    preds.append(pred)

c:\Users\Admin\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
c:\Users\Admin\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
c:\Users\Admin\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
c:\Users\Admin\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
c:\Users\Admin\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
c:\Users\Admin\anaconda3\

In [68]:
exp_sample

[0, 3.4, 7.5, 11.1, 10.5, 5, 0.5, 2.1]

In [69]:
print(preds)

[array([[24848.2]]), array([[56978.08]]), array([[95722.92]]), array([[129742.79]]), array([[124072.81]]), array([[72098.02]]), array([[29573.19]]), array([[44693.12]])]
